web/src/main/java/org/springframework/security/web/csrf/SecureCsrfToken.java

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright 2002-2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.security.web.csrf;
+
+import java.nio.ByteBuffer;
+import java.security.MessageDigest;
+import java.security.SecureRandom;
+import java.util.Base64;
+import org.springframework.security.crypto.codec.Utf8;
+import org.springframework.util.Assert;
+
+/**
+ * A CSRF token that is used to protect against CSRF attacks with extra protection from BREACH exploit
+ *
+ * @author Ruby Hartono
+ * @since 5.4
+ */
+@SuppressWarnings("serial")
+public final class SecureCsrfToken implements CsrfToken {
+
+	private final byte[] tokenBytes;
+
+	private final String parameterName;
+
+	private final String headerName;
+
+	private final SecureRandom secureRandom;
+
+	private final int randomByteSize;
+
+	/**
+	 * Creates a new instance
+	 *
+	 * @param headerName    the HTTP header name to use
+	 * @param parameterName the HTTP parameter name to use
+	 * @param token         the value of the token (i.e. expected value of the HTTP
+	 *                      parameter of parametername).
+	 */
+	public SecureCsrfToken(String headerName, String parameterName, String token) {
+		Assert.hasLength(headerName, "headerName cannot be null or empty");
+		Assert.hasLength(parameterName, "parameterName cannot be null or empty");
+		Assert.hasLength(token, "token cannot be null or empty");
+		this.headerName = headerName;
+		this.parameterName = parameterName;
+		this.tokenBytes = Utf8.encode(token);
+		this.secureRandom = new SecureRandom();
```

Comment:
There will be lots of instances of this object that are created. That means we should be able to inject and reuse the `SecureRandom` instance.